### PR TITLE
fix: prioritise error events over timeouts

### PR DESCRIPTION
**Pull request by jodevsa**
fix: prioritise error events over timeouts

Error events in the implementation of `net.connect` are defered to the next tick: https://github.com/nodejs/node/blob/main/lib/net.js#L1144
This will result in the TO callback always being executed before the error events callback in the callback queue.

This fix makes sure that error event callbacks are executed before the TO socket destroy logic.

fixes https://github.com/nodejs/undici/issues/1484

### CODE REVIEW

File: test/connect-timeout.js [hunk 1]
```diff
@@ -4,7 +4,31 @@ const { test } = require('tap')
...
 const net = require('net')
 
-// Never connect
+function blockingSleep (ms) {
+  let current = Date.now()
+  const blockUntil = current + ms
+  while (current < blockUntil) {
+    current = Date.now()
+  }
+}
```
Comment (mcollina) [inline test/connect-timeout.js:13]: Use https://www.npmjs.com/package/atomic-sleep instead

**Review by mcollina — COMMENTED** [review]
Good spot!

**Comment by jodevsa** [comment]
wasted 1 hour trying to find out why the test isn't working and realised that starbucks DNS resolves any hostname.!!
Screenshot 2022-07-16 at 20 22 05

**Comment by jodevsa** [comment]
@mcollina  please check again

File: lib/core/connect.js [hunk 2]
```diff
@@ -76,7 +76,7 @@ function buildConnector ({ maxCachedSessions, socketPath, timeout, ...opts }) {
     }
 
     const timeoutId = timeout
-      ? setTimeout(onConnectTimeout, timeout, socket)
+      ? setTimeout(() => setImmediate(() => onConnectTimeout(socket)), timeout)
```
Comment (ronag) [inline lib/core/connect.js:79]: I'm a little unsure about this... but at least we need a comment here explaining why.

Comment (jodevsa) [reply lib/core/connect.js:79]: sure, how about something like 

"setImmediate is needed to make sure that error event callbacks are executed before the TO socket destroy logic."?

**Comment by jodevsa** [comment]
tests are not working for windows

`Transformation error for D:\a\undici\undici\lib\core\connect.js ; return original code
EPERM: operation not permitted, rename 'D:\a\undici\undici\node_modules\.cache\nyc\fbfece793b45952b8a9635dc1bc6687ac6fb676105e3851ee22696fdcb8c0a14.js.349094085' -> 'D:\a\undici\undici\node_modules\.cache\nyc\fbfece793b45952b8a9635dc1bc6687ac6fb676105e3851ee22696fdcb8c0a14.js'
Transformation error for D:\a\undici\undici\lib\mock\mock-symbols.js ; return original code`

not sure what this is about

**Comment by jodevsa** [comment]
this change also seems to fix the example from the original issue

`const { Pool } = require('./')

const agent = new Pool('http://foobar.bar')

for (let i = 0; i < 50000; i++) {
  agent.request({ method: 'GET', path: '/foobar' })
    .then(async ({ body }) => {
      for await (const data of body) {
        console.log('data', data.toString('utf8'))
      }
    })
    .catch(console.error)
}
`

File: lib/core/connect.js (hunk 2 above)
Comment (jodevsa) [reply lib/core/connect.js:79]: > I'm a little unsure about this...
I was also not very sure about this but thinking of it again: Deferring throwing a TO in the hope that we already got a response makes sense.

**Comment by jodevsa** [comment]
my new test is failing on windows :(

**Comment by jodevsa** [comment]
@mcollina  seems like on windows we'll need 2 setImmediates I still don't know why but I noticed we have different implementation for windows in net.js. is it fine if we add another setImmediate?

**Comment by mcollina** [comment]
Detect if we are on windows and add another immediate in that case. I would also recommend moving that logic in a function because it's a bit hard to read.

**Comment by jodevsa** [comment]
@mcollina  please check again

**Comment by jodevsa** [comment]
It would be a good idea to cancel the setImmedates when the response/error is fired.. let me do that first and then I'll pinh you for revewing

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**